### PR TITLE
Removing CancelableAsyncEnumerable

### PR DESCRIPTION
**Pull request by ladeak**
**Removing CancelableAsyncEnumerable**
Using a while loop as discussed in a PR discussion previously: https://github.com/dotnet/aspnetcore/pull/6791#discussion_r259196518 

**Description**
- Using a while loop to iterate items
- Testing cancellation during streaming

Addresses https://github.com/dotnet/aspnetcore/issues/7960

### CODE REVIEW

**Comment by ladeak** [comment]
@BrennanConroy and @halter73 may I get a review?

File: src/SignalR/common/Shared/AsyncEnumerableAdapters.cs
Comment (halter73) [inline src/SignalR/common/Shared/AsyncEnumerableAdapters.cs:16]: If I recall this issue correctly, the idea was that we could avoid the allocation of a new `IAsyncEnumerable<object?>` altogether. This just changes it from a CancelableAsyncEnumerable allocation to an allocation of the [compiler-generated iterator](https://docs.microsoft.com/en-us/archive/msdn-magazine/2019/november/csharp-iterating-with-async-enumerables-in-csharp-8#under-the-hood-of-async-iterators) backing this async iterator method.

The idea behind #7960 is that we can just cast to `(IAsyncEnumerator<object?>)` without allocating at all if `!typeof(T).IsValueType` like I did [here in the original PR](https://github.com/dotnet/aspnetcore/pull/6791/commits/4de77e9c2c5f4cc9a18b1efcf4bf7906973861b1#diff-feae18e7b3265bffa80aab26f8dc8f8111684fda1e1b2781e775d5f69dc11851) before @davidfowl asked me to switch to a foreach and then switch it back after merging 😜.

Comment (ladeak) [reply src/SignalR/common/Shared/AsyncEnumerableAdapters.cs:16]: Got it, I will update the PR accordingly.

Comment (ladeak) [reply src/SignalR/common/Shared/AsyncEnumerableAdapters.cs:16]: So let me reiterate my current understanding. We would need to change `IAsyncEnumerable<object?> 
 MakeCancelableAsyncEnumerable` to `IAsyncEnumerator<object?> 
 MakeCancelableAsyncEnumerator` so by not having the enumerable but the enumerator, we avoid the allocation of the enumerable object for reference types.
For value types, we would still need to keep a `BoxedAsyncEnumerator`.

Because how `HubMethodDescriptor` uses these adapter methods, will need to update the `Channel<T>` signature and [implementation](https://github.com/dotnet/aspnetcore/blob/main/src/SignalR/common/Shared/AsyncEnumerableAdapters.cs#L26) too. If I see this correctly, for `Channel<T>` we will need to re-introduce a `ChannelAsyncEnumerator`  (like the one you had) probably for both netcore and netstandard implementation.

Then we would need to refactor [HubMethodDescriptor](https://github.com/dotnet/aspnetcore/blob/main/src/SignalR/server/Core/src/Internal/HubMethodDescriptor.cs#L121) to generate proper `Func<>` and then iterate using these iterators in [DefaultHubDispatcher](https://github.com/dotnet/aspnetcore/blob/main/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L478)

Is my understanding correct?
Is this the intended outcome the issue?

Comment (ladeak) [reply src/SignalR/common/Shared/AsyncEnumerableAdapters.cs:16]: @halter73 @BrennanConroy or @davidfowl could you advise on the above questions?

Comment (halter73) [reply src/SignalR/common/Shared/AsyncEnumerableAdapters.cs:16]: Your understanding sounds right to me! We'll need to keep a `BoxedAsyncEnumerator`. I think that a lot of the changes in https://github.com/dotnet/aspnetcore/pull/6791/commits/41544a289a99c9950b260beee329600b61dd8441 will need to be undone. Does that answer your questions?

Comment (ladeak) [reply src/SignalR/common/Shared/AsyncEnumerableAdapters.cs:16]: Yes, thank you.

Comment (ladeak) [reply src/SignalR/common/Shared/AsyncEnumerableAdapters.cs]: I have updated the PR as discussed.

File: src/SignalR/common/Shared/AsyncEnumerableAdapters.cs
Comment (halter73) [inline src/SignalR/common/Shared/AsyncEnumerableAdapters.cs:36]: Nit: we usually put properties below constructors and above methods.

File: src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs
Comment (halter73) [inline src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs:4578]: Nit: even in tests, we always use brackets for ifs and loops

File: src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs
Comment (halter73) [inline src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs:4594]: here too

File: src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs
Comment (halter73) [inline src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs:4598]:
```suggestion
```

File: src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs
Comment (halter73) [inline src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs:4153]: 👍 Nice test.

File: src/SignalR/common/Shared/AsyncEnumerableAdapters.cs
Comment (halter73) [inline src/SignalR/common/Shared/AsyncEnumerableAdapters.cs:65]: I think it would be better to call `Channel.TryRead()` in `MoveNextAsync()` and call `Channel.WaitToReadAsync()` (and then `Channel.TryRead()` again if necessary) here so we don't always need to observe the completion of this Channel with a `ChannelClosedException`.

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs
Comment (halter73) [inline src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs:473]:
```suggestion
                await using var enumerator = descriptor.FromReturnedStream(result, streamCts.Token);
```

File: src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs
Comment (halter73) [inline src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs:483]:
```suggestion
                        await connection.WriteAsync(streamItemMessage);
```

File: src/SignalR/common/Shared/AsyncEnumerableAdapters.cs
Comment (ladeak) [reply src/SignalR/common/Shared/AsyncEnumerableAdapters.cs:36]: Ack, I follow up on all comments added. Thanks for the review.

File: src/SignalR/common/Shared/AsyncEnumerableAdapters.cs
Comment (ladeak) [reply src/SignalR/common/Shared/AsyncEnumerableAdapters.cs]: ack

File: src/SignalR/common/Shared/AsyncEnumerableAdapters.cs
Comment (halter73) [inline src/SignalR/common/Shared/AsyncEnumerableAdapters.cs:55]: I did want to keep the `.ConfigureAwait(false)` in client code (and code in the `/Shared/` folder that could be used by the client).

```suggestion
                if (await _channel.WaitToReadAsync(_cancellationToken).ConfigureAwait(false) && _channel.TryRead(out var item))
```

**Review by halter73 — APPROVED** [review]
LGTM pending adding the `/Shared/` `.ConfigureAwait(false)` back. Thanks!!

File: src/SignalR/common/Shared/AsyncEnumerableAdapters.cs
Comment (ladeak) [reply src/SignalR/common/Shared/AsyncEnumerableAdapters.cs:55]: OK, I was actually wondering on that, makes sense now.

Comment (halter73) [reply src/SignalR/common/Shared/AsyncEnumerableAdapters.cs]: It was easy enough to commit from the github UI, so I did it for you. Thanks again!

**Comment by ladeak** [comment]
Thank you for the update. I am seeing that the Linux-ARM64 build is failing with

```
/home/vsts/work/1/s/eng/common/tools.sh: line 459:  2697 Segmentation fault      (core dumped) "$_InitializeBuildTool" "$@"
Build failed with exit code 139. Check errors above.
```

It seems this is some tooling issue. I would like to re-run builds (or at least that build) but I don't see a button for that. Is there a way to do that (other than pushing an update to the branch)?

**Review by BrennanConroy — APPROVED** [review]
Excellent! Thanks for doing this!

**Comment by BrennanConroy** [comment]
Thanks again @ladeak !

**Pull request merged**
